hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/node/SCMNodeManager.java

```diff
@@ -1642,4 +1642,34 @@ private ReentrantReadWriteLock.WriteLock writeLock() {
   private ReentrantReadWriteLock.ReadLock readLock() {
     return lock.readLock();
   }
+
+  /**
+   * This API allows removal of only DECOMMISSIONED and DEAD nodes from NodeManager data structures and cleanup memory.
+   * This API call is having a pre-condition before removal of node like following resources to be removed:
+   *   --- pipelines
+   *   --- containers
+   *   --- network topology
+   *   --- or any other cache related to node context in SCM
+   *
+   * @param datanodeDetails
+   * @throws NodeNotFoundException
+   */
+  @Override
+  public void removeNode(DatanodeDetails datanodeDetails) throws NodeNotFoundException, IOException {
+    writeLock().lock();
+    try {
+      NodeStatus nodeStatus = this.getNodeStatus(datanodeDetails);
+      if (datanodeDetails.isDecommissioned() || nodeStatus.isDead()) {
+        if (clusterMap.contains(datanodeDetails)) {
+          clusterMap.remove(datanodeDetails);
+        }
+        nodeStateManager.removeNode(datanodeDetails);
+        removeFromDnsToUuidMap(datanodeDetails.getUuid(), datanodeDetails.getIpAddress());
+        commandQueue.removeCommand(datanodeDetails.getUuid());
+        scmNodeEventPublisher.fireEvent(SCMEvents.DEAD_NODE, datanodeDetails);
```

Comment:
firing DEAD_NODE will try to get info from node manager and then cleanup, and to be called before removeNode. So IMO, firing this event i snot requried,

Reply:
For SCM, deadnode handle should be fired first -- this can be precondition.
For Recon, it sync details from SCM. So the state may change from SCM sync, do we really need this for this task?

Reply:
Node state sync doesn't happen from SCM in Recon, it is being updated based on DN heartbeat (stale and dead interval) and inherited DeadNodeHandler code of SCM gets called first in onMessage of ReconDeadNodeHandler, then recon updates the node operational state in ReconNodeManager class. I think , Recon don't sync node state from SCM.

Reply:
I have moved firing of DEAD_NODE event before  `nodeStateManager.removeNode(datanodeDetails);`

Reply:
Will remove DEAD_NODE fire event call and no needed, as this is async and cannot guarantee the cleanup before the node actually being removed. Also since this PR change will allow only decommissioned and DEAD nodes, so deadNodeHandler would have cleaned up already and decommissioning of node also destroy pipelines and cleanup container replicas.